File: src-tauri/src/storage.rs
use crate::models::{AppData, AppError, AppResult, AppState, CompoundRef, JobStatus, SharedState};
use log::{error, info, warn};
use std::path::{Path, PathBuf};
use tokio::task::JoinHandle;
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Load / persist state.json
// ---------------------------------------------------------------------------

/// Load state from `{root_dir}/state.json`, creating defaults if missing.
/// Also creates `state.json.bak` for crash recovery (D8).
pub fn load_state(root_dir: &Path) -> AppResult<AppState> {
    std::fs::create_dir_all(root_dir)?;

    let state_path = root_dir.join("state.json");

    let data = if state_path.exists() {
        // D8: backup before making any changes
        let bak_path = root_dir.join("state.json.bak");
        if let Err(e) = std::fs::copy(&state_path, &bak_path) {
            warn!("Failed to create state.json backup: {e}");
        }

        let content = std::fs::read_to_string(&state_path)?;
        serde_json::from_str(&content)?
    } else {
        AppData::default()
    };

    Ok(AppState {
        data,
        dirty: false,
        root_dir: root_dir.to_path_buf(),
    })
}

/// Atomic write: serialize → `.state.json.tmp` → rename.
/// Takes a cloned `AppData` so it can be called outside the lock (D1).
pub fn persist_state(root_dir: &Path, data: &AppData) -> AppResult<()> {
    let state_path = root_dir.join("state.json");
    let tmp_path = root_dir.join(".state.json.tmp");

    let content = serde_json::to_string_pretty(data)?;
    std::fs::write(&tmp_path, &content)?;
    std::fs::rename(&tmp_path, &state_path)?;

    Ok(())
}

/// D2: Spawns a 2-second interval flusher. If `dirty` is set, clones data,
/// resets the flag, drops the lock, then persists via spawn_blocking.
pub fn start_persistence_flusher(state: SharedState) -> JoinHandle<()> {
    tokio::spawn(async move {
        // A10: Skip the t=0 tick — start after the first 2-second delay
        let start = tokio::time::Instant::now() + std::time::Duration::from_secs(2);
        let mut interval = tokio::time::interval_at(start, std::time::Duration::from_secs(2));
        loop {
            interval.tick().await;

            let (should_persist, data_clone, root_dir) = {
                let mut guard = state.lock().await;
                if guard.dirty {
                    guard.dirty = false;
                    let clone = guard.data.clone();
                    let root = guard.root_dir.clone();
                    (true, Some(clone), Some(root))
                } else {
                    (false, None, None)
                }
            };

            if should_persist {
                if let (Some(data), Some(root)) = (data_clone, root_dir) {
                    // A10: Use spawn_blocking to avoid blocking the async executor
                    match tokio::task::spawn_blocking(move || persist_state(&root, &data)).await {
                        Ok(Err(e)) => error!("Persistence flusher failed: {e}"),
                        Err(e) => error!("Persistence flusher task panicked: {e}"),
                        _ => {}
                    }
                }
            }
        }
    })
}

// ---------------------------------------------------------------------------
// Folder operations
// ---------------------------------------------------------------------------

pub async fn create_campaign_folder(root: &Path, folder_name: &str) -> AppResult<()> {
    let path = root.join(folder_name);
    tokio::fs::create_dir_all(&path).await?;
    Ok(())
}

pub async fn create_run_folder(
    root: &Path,
    campaign_folder: &str,
    run_folder: &str,
) -> AppResult<()> {
    let path = root.join(campaign_folder).join(run_folder);
    tokio::fs::create_dir_all(&path).await?;
    Ok(())
}

/// A3: Validate that a folder name doesn't contain path traversal characters.
fn validate_folder_name(name: &str) -> AppResult<()> {
    if name.is_empty() || name.contains("..") || name.contains('/') || name.contains('\\') {
        return Err(AppError::Other(format!("Invalid folder name: {name}")));
    }
    Ok(())
}

/// Build the full path for a compound's output folder.
/// Returns a relative path from root_dir.
pub fn resolve_compound_path(data: &AppData, compound_id: Uuid) -> AppResult<PathBuf> {
    let (campaign, run, compound) = data
        .find_compound_context(compound_id)
        .ok_or_else(|| AppError::NotFound(format!("Compound {compound_id} not found")))?;

    // A3: Validate folder names to prevent path traversal from tampered state.json
    validate_folder_name(&campaign.folder_name)?;
    validate_folder_name(&run.folder_name)?;
    validate_folder_name(&compound.folder_name)?;

    Ok(PathBuf::from(&campaign.folder_name)
        .join(&run.folder_name)
        .join(&compound.folder_name))
}

pub async fn rename_folder(old: &Path, new: &Path) -> AppResult<()> {
    tokio::fs::rename(old, new).await?;
    Ok(())
}

/// Delete `.boltz-temp/` contents on startup.
pub async fn cleanup_temp_dir(root: &Path) -> AppResult<()> {
    let temp = root.join(".boltz-temp");
    if temp.exists() {
        info!("Cleaning up temp directory: {}", temp.display());
        tokio::fs::remove_dir_all(&temp).await?;
    }
    Ok(())
}

/// Find COMPLETED compounds that are missing their CIF files on disk.
pub fn scan_incomplete_downloads(root_dir: &Path, data: &AppData) -> Vec<CompoundRef> {
    let mut incomplete = Vec::new();
    for campaign in &data.campaigns {
        for run in &campaign.runs {
            for compound in &run.compounds {
                if compound.status == JobStatus::Completed {
                    if let (Some(job_id), Some(submitted_at)) =
                        (&compound.boltz_job_id, compound.submitted_at)
                    {
                        let compound_dir = root_dir
                            .join(&campaign.folder_name)
                            .join(&run.folder_name)
                            .join(&compound.folder_name);
                        let cif_path = compound_dir.join("sample_0_structure.cif");
                        if !cif_path.exists() {
                            incomplete.push(CompoundRef {
                                compound_id: compound.id,
                                boltz_job_id: job_id.clone(),
                                campaign_id: campaign.id,
                                run_id: run.id,
                                submitted_at,
                            });
                        }
                    }
                }
            }
        }
    }
    incomplete
}

File: src/main/services/storage.ts
import fs from 'node:fs';
import path from 'node:path';
import type { AppData, CompoundRef } from '../models/types';
import { AppState } from '../models/state';
import { FLUSH_INTERVAL_MS } from '../models/types';

// ── Load / Persist ──────────────────────────────────────────────────

/** Load state from {rootDir}/state.json, creating defaults if missing. */
export function loadState(rootDir: string): AppState {
  fs.mkdirSync(rootDir, { recursive: true });

  const statePath = path.join(rootDir, 'state.json');
  const backupPath = path.join(rootDir, 'state.json.bak');

  if (!fs.existsSync(statePath)) {
    return new AppState(AppState.defaultData(), rootDir);
  }

  const raw = fs.readFileSync(statePath, 'utf-8');
  const data: AppData = JSON.parse(raw);

  if (data.schema_version > 3) {
    throw new Error(
      `Unsupported state schema version: ${data.schema_version}. Please update Multiplexer.`,
    );
  }

  // Migrate v1 → v2: null out placeholder zero-affinity metrics
  let migrated = false;
  if (!data.schema_version || data.schema_version < 2) {
    for (const campaign of data.campaigns) {
      for (const run of campaign.runs) {
        for (const compound of run.compounds) {
          const aff = compound.metrics?.affinity;
          if (
            aff &&
            aff.binding_confidence === 0 &&
            aff.optimization_score === 0
          ) {
            compound.metrics!.affinity = null;
          }
        }
      }
    }
    data.schema_version = 2;
    migrated = true;
  }

  // Migrate v2 → v3: add target_type, rename protein_sequence → target_sequence
  if (data.schema_version < 3) {
    for (const campaign of data.campaigns) {
      const c = campaign as Record<string, unknown>;
      if (!c.target_type) {
        c.target_type = 'protein';
      }
      if ('protein_sequence' in c && !('target_sequence' in c)) {
        c.target_sequence = c.protein_sequence;
        delete c.protein_sequence;
      }
    }
    data.schema_version = 3;
    migrated = true;
  }

  // Create backup for crash recovery
  try {
    fs.copyFileSync(statePath, backupPath);
  } catch {
    // Non-fatal — backup creation is best-effort
  }

  const state = new AppState(data, rootDir);
  if (migrated) state.markDirty();
  return state;
}

/**
 * Atomic write: serialize → .state.json.tmp → rename.
 * Crash-safe on APFS (same-volume rename is atomic).
 */
export function persistState(rootDir: string, data: AppData): void {
  const statePath = path.join(rootDir, 'state.json');
  const tmpPath = path.join(rootDir, '.state.json.tmp');

  const json = JSON.stringify(data, null, 2);
  fs.writeFileSync(tmpPath, json, 'utf-8');
  fs.renameSync(tmpPath, statePath);
}

// ── Dirty-Flag Flusher ──────────────────────────────────────────────

/**
 * Start a 2-second interval that persists state when dirty.
 * Returns a cleanup function to stop the flusher.
 */
export function startPersistenceFlusher(state: AppState): () => void {
  const timer = setInterval(() => {
    if (!state.dirty) return;

    // Clone data, reset flag, then persist (I/O outside the "lock")
    const dataClone = structuredClone(state.data);
    state.dirty = false;
    persistState(state.rootDir, dataClone);
  }, FLUSH_INTERVAL_MS);

  return () => clearInterval(timer);
}

// ── Folder Operations ───────────────────────────────────────────────

export function createCampaignFolder(
  rootDir: string,
  folderName: string,
): void {
  validateFolderName(folderName);
  const dir = path.join(rootDir, folderName);
  fs.mkdirSync(dir, { recursive: true });
}

export function createRunFolder(
  rootDir: string,
  campaignFolder: string,
  runFolder: string,
): void {
  validateFolderName(campaignFolder);
  validateFolderName(runFolder);
  const dir = path.join(rootDir, campaignFolder, runFolder);
  fs.mkdirSync(dir, { recursive: true });
}

export function createCompoundFolder(
  rootDir: string,
  campaignFolder: string,
  runFolder: string,
  compoundFolder: string,
): void {
  validateFolderName(campaignFolder);
  validateFolderName(runFolder);
  validateFolderName(compoundFolder);
  const dir = path.join(rootDir, campaignFolder, runFolder, compoundFolder);
  fs.mkdirSync(dir, { recursive: true });
}

export function renameFolder(oldPath: string, newPath: string): void {
  fs.renameSync(oldPath, newPath);
}

export function cleanupTempDir(rootDir: string): void {
  const tempDir = path.join(rootDir, '.boltz-temp');
  if (fs.existsSync(tempDir)) {
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
}

// ── Path Resolution ─────────────────────────────────────────────────

export function resolveCompoundPath(
  state: AppState,
  compoundId: string,
): string {
  const ctx = state.findCompoundContext(compoundId);
  if (!ctx) throw new Error(`Compound not found: ${compoundId}`);
  const [campaign, run, compound] = ctx;
  return path.join(
    state.rootDir,
    campaign.folder_name,
    run.folder_name,
    compound.folder_name,
  );
}

// ── Validation ──────────────────────────────────────────────────────

function validateFolderName(name: string): void {
  if (!name || name.includes('..') || name.includes('/') || name.includes('\\')) {
    throw new Error(`Invalid folder name: "${name}"`);
  }
}

/**
 * Sanitise user-provided name into filesystem-safe folder name.
 * - Converts to lowercase
 * - Replaces non-alphanumeric chars with '-'
 * - Trims trailing '-'
 * - Truncates to ~200 bytes (UTF-8 safe)
 * - Returns "unnamed" if empty
 */
export function sanitiseFolderName(name: string): string {
  let result = name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/-+$/, '')
    .replace(/^-+/, '');

  // Truncate to ~200 chars
  if (result.length > 200) {
    result = result.slice(0, 200).replace(/-+$/, '');
  }

  return result || 'unnamed';
}

/**
 * Generate a unique folder name by appending -2, -3, etc. on collision.
 */
export function uniqueFolderName(
  baseName: string,
  existingNames: string[],
): string {
  const existing = new Set(existingNames);
  if (!existing.has(baseName)) return baseName;

  let counter = 2;
  while (existing.has(`${baseName}-${counter}`)) {
    counter++;
  }
  return `${baseName}-${counter}`;
}

// ── Download Recovery ───────────────────────────────────────────────

/**
 * Find COMPLETED compounds missing their CIF files on disk.
 * Used on startup before poller begins.
 */
export function scanIncompleteDownloads(
  rootDir: string,
  data: AppData,
): CompoundRef[] {
  const incomplete: CompoundRef[] = [];

  for (const campaign of data.campaigns) {
    for (const run of campaign.runs) {
      for (const compound of run.compounds) {
        if (compound.status !== 'COMPLETED') continue;
        if (!compound.boltz_job_id || !compound.submitted_at) continue;

        const cifPath = path.join(
          rootDir,
          campaign.folder_name,
          run.folder_name,
          compound.folder_name,
          'sample_0_structure.cif',
        );

        if (!fs.existsSync(cifPath)) {
          incomplete.push({
            compound_id: compound.id,
            boltz_job_id: compound.boltz_job_id,
            campaign_id: campaign.id,
            run_id: run.id,
            submitted_at: compound.submitted_at,
          });
        }
      }
    }
  }

  return incomplete;
}
